### DISCUSSION TITLE
Automate the process of passing the `create2Salt`

### DISCUSSION THREAD

**Discussion by andreivladbrg** [sablier-labs/evm-monorepo#810] [Ideas]
The current version of the deterministic deployment scripts requires manually passing the `create2Salt` which is a slow procces and prone to errors:

https://github.com/sablier-labs/v2-core/blob/7819a3b5122036596edff3326b97a8c03750bbb8/script/DeployDeterministicCore.s.sol#L22-L43

To automate this process, I've developed a utility that builds the salt using the [ffi](https://book.getfoundry.sh/cheatcodes/ffi?highlight=ffi#ffi) cheatcode, which allows you to execute an arbitrary shell command and capture the output: 

```solidity
string memory chainId = block.chainid.toString();
string[] memory inputs = new string[](4);
inputs[0] = "jq";
inputs[1] = "-r";
inputs[2] = ".version";
inputs[3] = "./package.json";
bytes memory result = vm.ffi(inputs);
string memory version = string(result);
string memory create2Salt = string.concat("ChainID ", chainId, ", Version ", version);
```
**Note:** it would require to import and use a library to convert uint256 to string

I've personally tested this and it works. https://app.warp.dev/block/5564ry0K2Lxbq9seHWJKsx

The most appropriate place to implement this is in a function named `getCreate2Salt` within `Base_Script`.

wdyt @PaulRBerg @smol-ninja ?
I believe you are going to like it 😁

**Comment by smol-ninja** [comment]
Nice! I love your approach of `ffi`. However, comparing with the current implementation, I do not completely agree with the statement:

> manually passing the `create2Salt`  which is a slow process and prone to errors

While it might have been prone to errors in the past, we've since resolved those issues. Right now, it simply involves passing the version ID (manually), albeit at a slightly slower pace compared to utilizing `ffi`.

<img width="800" alt="Screenshot 2024-02-01 at 18 26 25" src="https://github.com/sablier-labs/v2-core/assets/6676622/04e16e22-339e-4ff6-aae9-69a718c85dd4">


Additionally, what if, in the future, we consider to change to the salt structure? My interpretation is that a salt is intended to be a source of external randomness. As such, I still think we should keep it outside the foundry to maintain flexibility in case any adjustments are needed down the line.

**Reply by andreivladbrg** [reply]
> While it might have been prone to errors in the past

it still is

>it simply involves passing the version ID (manually)

doing something manually  ⇒ prone to errors

also, we would not even need to declare manually [the chain ids](https://github.com/sablier-labs/v2-core/blob/84758a40077bf3ccb1c8f7bb8d00278e672fbfef/shell/deploy-multi-chain.sh#L89-L100)

**Reply by smol-ninja** [reply]
> we would not even need to declare manually [the chain ids](https://github.com/sablier-labs/v2-core/blob/84758a40077bf3ccb1c8f7bb8d00278e672fbfef/shell/deploy-multi-chain.sh#L89-L100)

That's a good point. Less number of lines in the bash script ftw 🎉.

> doing something manually ⇒ prone to errors

OK, but any additions to the [.env.deployment](https://github.com/sablier-labs/v2-core/blob/84758a40077bf3ccb1c8f7bb8d00278e672fbfef/.env.deployment.example) also involve manual intervention. We can also extract `version` from `package.json` in the bash script itself. But I agree that it would still require us to have chain id manually added in the script which is an advantage that `ffi` has.

**Reply by PaulRBerg** [reply]
> we've since resolved those issues. Right now, it simply involves passing the version ID (manually)

Only if using the `deploy-multi-chain` script.

> My interpretation is that a salt is intended to be a source of external randomness

Quite the opposite - we use it to provide [deterministic differentiation](https://twitter.com/PaulRBerg/status/1677610313993146368).

> in case any adjustments are needed down the line

For rapid deployments, we can use the non-deterministic deployment scripts.

**Reply by smol-ninja** [reply]
Agree to all your points.

**Comment by PaulRBerg** [comment]
I like the proposal a lot, @andreivladbrg. Feel free to go ahead with implementing it.
